### ISSUE TITLE
Unable to Connect to SoulSeek Server when Dual Network Interface Is Configured

### ISSUE THREAD

**Issue by frnksmdlkedjnfr** [nicotine-plus/nicotine-plus#2518]
Nicotine+ version: 3.2.8  •  GTK 3.24.36
Operating System/Distribution: Windows 11 Pro for Workstations 22H2 22621.1778

**### Describe the bug**
My company isolates the Internet from the Intranet, so I had to write some extra route commands to access the Internet. However, while 99% of applications and websites work fine, Nicotine+ is an exception. It just won't connect to the SoulSeek server. No system proxy is being using. The following GIF describes the issue I'm having:
![录制_2023_05_29_09_29_36_278](https://github.com/nicotine-plus/nicotine-plus/assets/63704404/23692215-e205-4fd8-8bd7-ec00602f080a)

tracert looks normal:
![2023-05-29_09-53-22_Administrator Windows Pow
![2023-05-29_09-53-22_Administrator Windows PowerShell](https://github.com/nicotine-plus/nicotine-plus/assets/63704404/afc6add3-e39e-436f-8a3e-2d5b451f6447)
erShell](https://github.com/nicotine-plus/nicotine-plus/assets/63704404/98203b8d-98c9-4280-bb83-5a7eb7618f11)

**### Expected behavior**
![2023-05-29_09-59-06_Nicotine+](https://github.com/nicotine-plus/nicotine-plus/assets/63704404/730d8642-d351-4089-95b1-9143a1b8e649)

**### Steps to reproduce the bug**

1. Plug in the LAN cable;
2. Wait for the Ethernet to identify;
3. Execute the following CMD command to configure the route table:

`route delete 0.0.0.0 mask 0.0.0.0 10.134.48.1 metric 35 -p`
`route delete 10.0.0.0 mask 255.0.0.0 10.134.48.1 metric 50`

**### Additional context**
My route table:
![2023-05-29_10-05-01_Administrator Windows PowerShell](https://github.com/nicotine-plus/nicotine-plus/assets/63704404/3a6b8aef-3101-41fa-81b5-4f0e571378a5)

**Screenshots, logs, stacktraces or relevant information.**
1. 10 out of 10 times, unplug the LAN cable and Nicotine+ instantly connects;
2. With LAN cable connected, after 15 minutes or so, Nicotine+ would randomly connects;
3. My coworker from the IT department suspects that Nicotine+ isn't optimised for dual NIC routes.

**Comment by slook** [comment]
What program is that firewall connection manager app in your tray called?

**Comment by frnksmdlkedjnfr** [comment]
That programme is called Clash Verge. It's a VPN. I showed in the gif that I was not using any VPN at that time.

And as for `route delete 0.0.0.0 mask 0.0.0.0 10.134.48.1 metric 35 -p`, this one is the redundant route created by the LAN network. If that remains, it would mess the route table up, causing webpages to load forever etc. So I had to delete that redundant route.

Also you can check my route table. Basically there's two: the 0.0.0.0 is for Internet, 10.0.0.0 is for Intranet. For any other programme and website they work fine.

**Comment by slook** [comment]
You might need #871 , currently Nicotine+ supports network interface binding on all OS's other than Windows.

Do you have access to any Linux or macOS machines on your network?

**Comment by frnksmdlkedjnfr** [comment]
I'm not that techy so I had trouble grasping what that issue was discussing. However I do run a Linux (WSL 2) virtual machine on my Windows laptop. Don't know if that'll do.

**Comment by slook** [comment]
> I'm not that techy 

I disagree, you seem to have a good understanding of your complex network topology.

It would be useful to know if the bug is specific to your OS or if it has to do with your network environment. I.e. if binding to a specific network interface solves your issue, then that might help us to determine either that there is a bug that needs fixing, or simply a required function that is lacking.

I'm not sure if a virtual WSL environment would be suitable for narrowing down the issue or not, that would depend upon how the network interface(s) is/are exposed to the OS running on the machine. Therefore it's not worth you doing any significant effort to try this.

We don't have many Windows testers, so it might take a while until this issue can be further investigated.

Do you experience the same problem with the [official SoulseekQt client](https://www.slsknet.org/news/node/1)?

**Comment by frnksmdlkedjnfr** [comment]
Unfortunately I don't have a device physically running MacOS or Linux available at hand. But as for the official SSQT client. It seems to have the same problem. But it's been long since I last used it.

Nicotine+ turned out find in the past few days. It's probably some VPN interference.

And as you mentioned the paucity of Windows testers, would I be able to opt in your beta programme (if any)?

**Comment by slook** [comment]
> It's probably some VPN interference.

There are some proprietary routers and network appliances which seem to block potentially unwanted programs and services that create the type of P2P traffic patterns that slsk users create. If the problem is intermittent then this is likely the case here.

> the paucity of Windows testers, would I be able to opt in your beta programme (if any)?

Sure! That would be much appreciated... There's nothing formal, the unstable builds install in exactly the same way as the normal releases, see link: https://github.com/nicotine-plus/nicotine-plus/blob/master/doc/TESTING.md#windows

At the moment we're working on Nicotine+ 3.3.0.dev5 due for release later this year. Bug reports are always welcome on here.

**Comment by frnksmdlkedjnfr** [comment]
So I further tested this problem today. Demonstration is shown in the gif:
![录制_2023_06_02_16_06_31_969](https://github.com/nicotine-plus/nicotine-plus/assets/63704404/46dfba12-e7c7-4ab8-8e07-1676341d895b)

As you can see, with dual network interfaces (probably) properly configured with routing rules, Nicotine+ (in most cases) cannot connect to the server. However, when I turn on the TUN mode of Clash Verge (VPN I'm using), Nicotine+ suddenly was able to connect.

The TUN mode in Clash basically creates a virtual network card, and forces all traffic through this interface. So shown in the gif, the route table is changed.

I assume that Nicotine+ might has its own network routing rules, so it doesn't use normal HTTP proxy or PAC proxy. This happens to some games like COD Cold War 2.

The reason why I said in most cases Nicotine+ cannot connect to the server, is actually a fact. Sometimes even without TUN mode on, Nicotine+ would happen to connect to the server. And as far as I have observed, so long as you had Nicotine+ connected to the server, the connection won't stop when you had the route table configured right as plugging the LAN cable (I guess it has something to do with the heartbeat span between Nicotine+ and the server. I'm not familiar with this stuff).

**Comment by slook** [comment]
@frnksmdlkedjnfr You will be interested to try out the new network interface binding feature that has been recently added for Windows in Nicotine+ 3.3.0.dev5...

https://github.com/nicotine-plus/nicotine-plus/blob/master/doc/TESTING.md#windows

Menu > Preferences > Network > Network interface

It might not solve your issue, but it will help us to rule out if the problem was the lacking bind feature or not.

**Comment by frnksmdlkedjnfr** [comment]
Thx for mentioning me. I tried this dev build and it had no problem
connecting to the server no matter what network interface I choose.

One more thing tho, will auto-dark (using system settings) mode be
available anytime soon?

**Comment by mathiascode** [comment]
> One more thing tho, will auto-dark (using system settings) mode be available anytime soon?

That should already be available in 3.3.0.dev5. Does it not work on your system?

**Comment by mathiascode** [comment]
Closing as resolved, let me know if you still experience the issue.
